### PR TITLE
TMEDIA-194 - Ad Block reserve space

### PR DESCRIPTION
**Pull request by matthewroach**
## Description

Add new custom field to ad-block which defaults to true - which ads logic to reserve space for the ad to fill to avoid layout shifting.

## Jira Ticket
- [TMEDIA-194](https://arcpublishing.atlassian.net/browse/TMEDIA-194)

## Acceptance Criteria
1. A new boolean Custom Field is added to the Ad Block called Reserve space for ad? 
    * If the value is true then the block should reserve enough space for the ad to fill in on client side load
    * If the value is false or not present then no change in behavior needs to happen, it behaves as it does now (not reserving space) 
    * The field should default to true
2. If there are multiple heights requested for the ad, it should reserve the smaller height. It should reserve the relevant height by breakpoint.

## Test Steps

1. Checkout this branch `git checkout TMEDIA-194-reserve-ad-space`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/ads-block`
3. Visit a page with ads on and verify page loads with space for the ad to fill, and then ads are loaded and do not cause layout shifting - http://localhost/pf/adstesting/?_website=the-gazette
4. Article page with Ads - http://localhost/2019/09/17/global-kitchen-sink-article/?_website=the-gazette
5. Edit the ads testing page to disabled reserve space and verify layout shifting is back - Not custom fields for existing blocks on a page do inherit default values.

### Before

Ads caused lot of layout shifting

TMEDIA-194-before

### After

Reduced layout shifting - still see some layout shift based on what ad is being served in slot that have multiple sizes allocated to them

TMEDIA-194-after

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [X] Confirmed all the test steps a reviewer will follow above are working.
- [X] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [X] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [X] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [X] Confirmed this PR has unit test files
  - [X] Ran `npm run test`, made sure all tests are passing
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

### CODE REVIEW

File: blocks/ads-block/features/ads/default.jsx [hunk 1]
```diff
@@ -33,7 +33,6 @@ const StyledAdUnit = styled.div`
 `;
 
 const ArcAd = (props) => {
-  if (typeof window === 'undefined') return null;
```
Comment (JackHowa) [inline blocks/ads-block/features/ads/default.jsx:36]: was this change deliberate?

File: blocks/ads-block/features/ads/default.test.jsx [hunk 2]
```diff
@@ -96,14 +97,38 @@ describe('<ArcAd>', () => {
...
-        const adUnitEl = wrapper.find('.arcad_feature LazyLoad AdUnit');
+        const adUnitEl = wrapper.find('.arcad-feature LazyLoad AdUnit');
         expect(adUnitEl).toHaveLength(1);
         expect(typeof adUnitEl.prop('adConfig')).toEqual('object');
         expect(typeof adUnitEl.prop('featureConfig')).toEqual('object');
       });
     });
   });
 
+  describe('Reserve Space', () => {
```
Comment (JackHowa) [inline blocks/ads-block/features/ads/default.test.jsx:108]: looks good, nice work

File: blocks/shared-styles/_children/promo-date/index.test.jsx [hunk 3]
```diff
@@ -15,7 +15,7 @@ jest.mock('fusion:properties', () => jest.fn(() => (
 )));
 
 jest.mock('@wpmedia/engine-theme-sdk', () => ({
-  localizeDateTime: jest.fn(() => new Date().toDateString()),
+  localizeDateTime: jest.fn((x) => new Date(x).toDateString()),
```
Comment (JackHowa) [inline blocks/shared-styles/_children/promo-date/index.test.jsx:18]: ew, I'm not sure how I feel about this mock. good job improving it though. I might prefer just to pass through a fake date?

**Review by JackHowa — DISMISSED** [review]
I think we need to keep the window check if I'm not mistaken. the rest looks pretty solid, especially with min-content

File: blocks/ads-block/features/ads/default.jsx (hunk 1 above)
Comment (matthewroach) [reply blocks/ads-block/features/ads/default.jsx:36]: Needs to return from the server so we can place the element on the page to reserve the space

Comment (JackHowa) [reply blocks/ads-block/features/ads/default.jsx:36]: wait, we need this

Comment (JackHowa) [reply blocks/ads-block/features/ads/default.jsx:36]: yep just missed your comment. was thinking about this more heh

Comment (vinnie-olsen) [reply blocks/ads-block/features/ads/default.jsx:36]: Just for context, I originally added the window check to prevent the ads code from executing server-side and rendering a blank ad placeholder element. That being said, I was planning on refactoring this because I think it breaks the rules of hooks since it returns before the `useState` calls (but then I left the team and never got a chance to actually do that). 

I don't _think_ removing it should cause any issues since I don't think the `arcads` library should actually make ad requests when server-side. I'm not 100% sure on that part though. There might need to be a check to prevent `arcads` from being initialized & registering ad units when server-side if not preventing the `ads-block` from rendering server-side. Not completely sure without testing that though.

Comment (JackHowa) [reply blocks/ads-block/features/ads/default.jsx:36]: ok cool, we'll move forward with removing this. @matthewroach can you remove the rule of hooks ignore in the first line

**Review by JackHowa — DISMISSED** [review]

File: blocks/ads-block/features/ads/default.jsx (hunk 1 above)
Comment (JackHowa) [reply blocks/ads-block/features/ads/default.jsx:36]: thanks for the clarification @vinnie-olsen !

**Review by JackHowa — APPROVED** [review]

File: blocks/ads-block/features/ads/default.jsx (hunk 1 above)
Comment (vinnie-olsen) [reply blocks/ads-block/features/ads/default.jsx:36]: no problem @JackHowa 🙂
